File: ios/runanywhere-sdks-latest/sdk/runanywhere-react-native/packages/core/ios/ArchiveUtility.swift
/**
 * ArchiveUtility.swift
 *
 * Native archive extraction utility for React Native.
 * Uses Apple's native Compression framework for gzip decompression (fast)
 * and pure Swift tar extraction.
 *
 * Mirrors the implementation from:
 * sdk/runanywhere-swift/Sources/RunAnywhere/Infrastructure/Download/Utilities/ArchiveUtility.swift
 *
 * Supports: tar.gz, zip
 * Note: All models should use tar.gz from RunanywhereAI/sherpa-onnx fork for best performance
 */

import Compression
import Foundation

/// Archive extraction errors
public enum ArchiveError: Error, LocalizedError {
    case invalidArchive(String)
    case decompressionFailed(String)
    case extractionFailed(String)
    case unsupportedFormat(String)
    case fileNotFound(String)

    public var errorDescription: String? {
        switch self {
        case .invalidArchive(let msg): return "Invalid archive: \(msg)"
        case .decompressionFailed(let msg): return "Decompression failed: \(msg)"
        case .extractionFailed(let msg): return "Extraction failed: \(msg)"
        case .unsupportedFormat(let msg): return "Unsupported format: \(msg)"
        case .fileNotFound(let msg): return "File not found: \(msg)"
        }
    }
}

/// Utility for handling archive extraction
@objc public final class ArchiveUtility: NSObject {

    // MARK: - Public API

    /// Extract an archive to a destination directory
    /// - Parameters:
    ///   - archivePath: Path to the archive file
    ///   - destinationPath: Destination directory path
    /// - Returns: true if extraction succeeded
    @objc public static func extract(
        archivePath: String,
        to destinationPath: String
    ) -> Bool {
        do {
            try extractArchive(archivePath: archivePath, to: destinationPath)
            return true
        } catch {
            SDKLogger.archive.logError(error, additionalInfo: "Extraction failed")
            return false
        }
    }

    /// Extract an archive to a destination directory (throwing version)
    public static func extractArchive(
        archivePath: String,
        to destinationPath: String,
        progressHandler: ((Double) -> Void)? = nil
    ) throws {
        let archiveURL = URL(fileURLWithPath: archivePath)
        let destinationURL = URL(fileURLWithPath: destinationPath)

        // Ensure archive exists
        guard FileManager.default.fileExists(atPath: archivePath) else {
            throw ArchiveError.fileNotFound("Archive not found: \(archivePath)")
        }

        // Detect archive type by magic bytes (more reliable than file extension)
        let archiveType = try detectArchiveTypeByMagicBytes(archivePath)
        SDKLogger.archive.info("Detected archive type: \(archiveType) for: \(archivePath)")

        switch archiveType {
        case .gzip:
            try extractTarGz(from: archiveURL, to: destinationURL, progressHandler: progressHandler)
        case .zip:
            try extractZip(from: archiveURL, to: destinationURL, progressHandler: progressHandler)
        case .bzip2:
            throw ArchiveError.unsupportedFormat("tar.bz2 not supported. Use tar.gz from RunanywhereAI/sherpa-onnx fork.")
        case .xz:
            throw ArchiveError.unsupportedFormat("tar.xz not supported. Use tar.gz from RunanywhereAI/sherpa-onnx fork.")
        case .unknown:
            // Fallback to file extension check
            let lowercased = archivePath.lowercased()
            if lowercased.hasSuffix(".tar.gz") || lowercased.hasSuffix(".tgz") {
                try extractTarGz(from: archiveURL, to: destinationURL, progressHandler: progressHandler)
            } else if lowercased.hasSuffix(".zip") {
                try extractZip(from: archiveURL, to: destinationURL, progressHandler: progressHandler)
            } else {
                throw ArchiveError.unsupportedFormat("Unknown archive format: \(archivePath)")
            }
        }
    }

    /// Archive type detected by magic bytes
    private enum DetectedArchiveType {
        case gzip
        case zip
        case bzip2
        case xz
        case unknown
    }

    /// Detect archive type by reading magic bytes from file header
    private static func detectArchiveTypeByMagicBytes(_ path: String) throws -> DetectedArchiveType {
        guard let fileHandle = FileHandle(forReadingAtPath: path) else {
            throw ArchiveError.fileNotFound("Cannot open file: \(path)")
        }
        defer { try? fileHandle.close() }

        // Read first 6 bytes for magic number detection
        guard let headerData = try? fileHandle.read(upToCount: 6), headerData.count >= 2 else {
            return .unknown
        }

        // Check for gzip: 0x1f 0x8b
        if headerData[0] == 0x1f && headerData[1] == 0x8b {
            return .gzip
        }

        // Check for zip: 0x50 0x4b 0x03 0x04 ("PK\x03\x04")
        if headerData.count >= 4 &&
           headerData[0] == 0x50 && headerData[1] == 0x4b &&
           headerData[2] == 0x03 && headerData[3] == 0x04 {
            return .zip
        }

        // Check for bzip2: 0x42 0x5a ("BZ")
        if headerData[0] == 0x42 && headerData[1] == 0x5a {
            return .bzip2
        }

        // Check for xz: 0xfd 0x37 0x7a 0x58 0x5a 0x00
        if headerData.count >= 6 &&
           headerData[0] == 0xfd && headerData[1] == 0x37 &&
           headerData[2] == 0x7a && headerData[3] == 0x58 &&
           headerData[4] == 0x5a && headerData[5] == 0x00 {
            return .xz
        }

        return .unknown
    }

    // MARK: - tar.gz Extraction (Native Compression Framework)

    /// Extract a tar.gz archive using streaming decompression to keep memory constant.
    /// Decompresses gzip to a temporary tar file on disk, then extracts tar entries.
    private static func extractTarGz(
        from sourceURL: URL,
        to destinationURL: URL,
        progressHandler: ((Double) -> Void)?
    ) throws {
        let overallStart = Date()
        SDKLogger.archive.info("Extracting tar.gz: \(sourceURL.lastPathComponent)")
        progressHandler?(0.0)

        // Step 1: Stream-decompress gzip to a temporary tar file on disk.
        // This avoids holding both compressed + decompressed data in memory simultaneously.
        let tempTarURL = destinationURL.appendingPathExtension("tar.tmp")
        defer { try? FileManager.default.removeItem(at: tempTarURL) }

        let decompressStart = Date()
        SDKLogger.archive.info("Starting streaming gzip decompression...")
        try decompressGzipToFile(from: sourceURL, to: tempTarURL)
        let decompressTime = Date().timeIntervalSince(decompressStart)
        let tarFileSize = (try? FileManager.default.attributesOfItem(atPath: tempTarURL.path)[.size] as? Int) ?? 0
        SDKLogger.archive.info("Decompressed to \(formatBytes(tarFileSize)) in \(String(format: "%.2f", decompressTime))s")
        progressHandler?(0.3)

        // Step 2: Read tar data and extract entries
        let extractStart = Date()
        SDKLogger.archive.info("Extracting tar data...")
        let tarData = try Data(contentsOf: tempTarURL)
        try extractTarData(tarData, to: destinationURL, progressHandler: { progress in
            progressHandler?(0.3 + progress * 0.7)
        })
        let extractTime = Date().timeIntervalSince(extractStart)
        SDKLogger.archive.info("Tar extract completed in \(String(format: "%.2f", extractTime))s")

        let totalTime = Date().timeIntervalSince(overallStart)
        SDKLogger.archive.info("Total extraction time: \(String(format: "%.2f", totalTime))s")
        progressHandler?(1.0)
    }

    /// Stream-decompress a gzip file to an output file using compression_stream_process.
    /// Peak memory usage is ~512 KB (input + output buffers) regardless of file size.
    private static func decompressGzipToFile(from sourceURL: URL, to destinationURL: URL) throws {
        guard let inputHandle = FileHandle(forReadingAtPath: sourceURL.path) else {
            throw ArchiveError.fileNotFound("Cannot open: \(sourceURL.path)")
        }
        defer { try? inputHandle.close() }

        // Parse gzip header to find where the deflate stream begins
        let headerOffset = try parseGzipHeader(from: inputHandle)
        inputHandle.seek(toFileOffset: UInt64(headerOffset))

        // The deflate stream ends 8 bytes before EOF (CRC32 + ISIZE trailer)
        let attrs = try FileManager.default.attributesOfItem(atPath: sourceURL.path)
        let fileSize = (attrs[.size] as? UInt64) ?? 0
        guard fileSize > UInt64(headerOffset) + 8 else {
            throw ArchiveError.invalidArchive("Gzip file too small for valid deflate stream")
        }
        let deflateEndOffset = fileSize - 8

        // Create output file
        FileManager.default.createFile(atPath: destinationURL.path, contents: nil)
        guard let outputHandle = FileHandle(forWritingAtPath: destinationURL.path) else {
            throw ArchiveError.extractionFailed("Cannot create temp file: \(destinationURL.path)")
        }
        defer { try? outputHandle.close() }

        // Initialize streaming decompression
        let dummyStreamPointer = UnsafeMutablePointer<UInt8>.allocate(capacity: 1)
        defer { dummyStreamPointer.deallocate() }
        var stream = compression_stream(
            dst_ptr: dummyStreamPointer,
            dst_size: 0,
            src_ptr: UnsafePointer(dummyStreamPointer),
            src_size: 0,
            state: nil
        )
        guard compression_stream_init(&stream, COMPRESSION_STREAM_DECODE, COMPRESSION_ZLIB) == COMPRESSION_STATUS_OK else {
            throw ArchiveError.decompressionFailed("Failed to initialize decompression stream")
        }
        defer { compression_stream_destroy(&stream) }

        let chunkSize = 256 * 1024 // 256 KB
        let inputBuffer = UnsafeMutablePointer<UInt8>.allocate(capacity: chunkSize)
        let outputBuffer = UnsafeMutablePointer<UInt8>.allocate(capacity: chunkSize)
        defer {
            inputBuffer.deallocate()
            outputBuffer.deallocate()
        }

        stream.src_size = 0
        var finished = false

        while !finished {
            // Feed more input when the previous chunk is consumed
            if stream.src_size == 0 {
                let currentOffset = inputHandle.offsetInFile
                if currentOffset >= deflateEndOffset {
                    finished = true
                } else {
                    let bytesToRead = min(UInt64(chunkSize), deflateEndOffset - currentOffset)
                    let chunk = inputHandle.readData(ofLength: Int(bytesToRead))
                    if chunk.isEmpty {
                        finished = true
                    } else {
                        chunk.copyBytes(to: inputBuffer, count: chunk.count)
                        stream.src_ptr = UnsafePointer(inputBuffer)
                        stream.src_size = chunk.count
                    }
                }
            }

            stream.dst_ptr = outputBuffer
            stream.dst_size = chunkSize

            let flags: Int32 = finished ? Int32(COMPRESSION_STREAM_FINALIZE.rawValue) : 0
            let status = compression_stream_process(&stream, flags)

            let bytesProduced = chunkSize - stream.dst_size
            if bytesProduced > 0 {
                outputHandle.write(Data(bytes: outputBuffer, count: bytesProduced))
            }

            switch status {
            case COMPRESSION_STATUS_OK:
                continue
            case COMPRESSION_STATUS_END:
                finished = true
            case COMPRESSION_STATUS_ERROR:
                throw ArchiveError.decompressionFailed("Streaming decompression error")
            default:
                throw ArchiveError.decompressionFailed("Unexpected compression status: \(status)")
            }
        }
    }

    /// Parse the gzip header and return the byte offset where the deflate stream begins.
    private static func parseGzipHeader(from handle: FileHandle) throws -> Int {
        handle.seek(toFileOffset: 0)
        guard let header = try? handle.read(upToCount: 10), header.count >= 10 else {
            throw ArchiveError.invalidArchive("Gzip data too short")
        }
        guard header[0] == 0x1f && header[1] == 0x8b else {
            throw ArchiveError.invalidArchive("Invalid gzip magic number")
        }
        guard header[2] == 8 else {
            throw ArchiveError.invalidArchive("Unsupported gzip compression method")
        }

        let flags = header[3]
        var offset = 10

        if (flags & 0x04) != 0 { // FEXTRA
            handle.seek(toFileOffset: UInt64(offset))
            guard let extraLenData = try? handle.read(upToCount: 2), extraLenData.count >= 2 else {
                throw ArchiveError.invalidArchive("Truncated gzip header (FEXTRA)")
            }
            let extraLen = Int(extraLenData[0]) | (Int(extraLenData[1]) << 8)
            offset += 2 + extraLen
        }

        if (flags & 0x08) != 0 { // FNAME
            handle.seek(toFileOffset: UInt64(offset))
            while true {
                guard let byte = try? handle.read(upToCount: 1), byte.count == 1 else { break }
                offset += 1
                if byte[0] == 0 { break }
            }
        }

        if (flags & 0x10) != 0 { // FCOMMENT
            handle.seek(toFileOffset: UInt64(offset))
            while true {
                guard let byte = try? handle.read(upToCount: 1), byte.count == 1 else { break }
                offset += 1
                if byte[0] == 0 { break }
            }
        }

        if (flags & 0x02) != 0 { // FHCRC
            offset += 2
        }

        return offset
    }

    // MARK: - ZIP Extraction (Pure Swift using Foundation)

    private static func extractZip(
        from sourceURL: URL,
        to destinationURL: URL,
        progressHandler: ((Double) -> Void)?
    ) throws {
        SDKLogger.archive.info("Extracting zip: \(sourceURL.lastPathComponent)")
        progressHandler?(0.0)

        // Create destination directory
        try FileManager.default.createDirectory(at: destinationURL, withIntermediateDirectories: true)

        // Read zip file
        guard let archive = try? Data(contentsOf: sourceURL) else {
            throw ArchiveError.fileNotFound("Cannot read zip file: \(sourceURL.path)")
        }

        // Parse and extract ZIP using pure Swift
        var offset = 0
        var fileCount = 0
        let totalSize = archive.count

        while offset < archive.count - 4 {
            // Check for local file header signature (0x04034b50 = PK\x03\x04)
            let sig0 = archive[offset]
            let sig1 = archive[offset + 1]
            let sig2 = archive[offset + 2]
            let sig3 = archive[offset + 3]

            if sig0 == 0x50 && sig1 == 0x4b && sig2 == 0x03 && sig3 == 0x04 {
                // Local file header
                let compressionMethod = UInt16(archive[offset + 8]) | (UInt16(archive[offset + 9]) << 8)
                let compressedSize = UInt32(archive[offset + 18]) |
                    (UInt32(archive[offset + 19]) << 8) |
                    (UInt32(archive[offset + 20]) << 16) |
                    (UInt32(archive[offset + 21]) << 24)
                let uncompressedSize = UInt32(archive[offset + 22]) |
                    (UInt32(archive[offset + 23]) << 8) |
                    (UInt32(archive[offset + 24]) << 16) |
                    (UInt32(archive[offset + 25]) << 24)
                let fileNameLength = UInt16(archive[offset + 26]) | (UInt16(archive[offset + 27]) << 8)
                let extraFieldLength = UInt16(archive[offset + 28]) | (UInt16(archive[offset + 29]) << 8)

                let headerEnd = offset + 30
                let fileNameData = archive.subdata(in: headerEnd..<(headerEnd + Int(fileNameLength)))
                let fileName = String(data: fileNameData, encoding: .utf8) ?? ""

                let dataStart = headerEnd + Int(fileNameLength) + Int(extraFieldLength)
                let dataEnd = dataStart + Int(compressedSize)

                let filePath = destinationURL.appendingPathComponent(fileName)

                if fileName.hasSuffix("/") {
                    // Directory
                    try FileManager.default.createDirectory(at: filePath, withIntermediateDirectories: true)
                } else if !fileName.isEmpty && !fileName.hasPrefix("__MACOSX") {
                    // File
                    try FileManager.default.createDirectory(at: filePath.deletingLastPathComponent(), withIntermediateDirectories: true)

                    if compressionMethod == 0 {
                        // Stored (no compression)
                        let fileData = archive.subdata(in: dataStart..<dataEnd)
                        try fileData.write(to: filePath)
                    } else if compressionMethod == 8 {
                        // Deflate compression - use native decompression
                        let compressedData = archive.subdata(in: dataStart..<dataEnd)
                        if let decompressed = decompressDeflateDataForZip(compressedData, uncompressedSize: Int(uncompressedSize)) {
                            try decompressed.write(to: filePath)
                        } else {
                            SDKLogger.archive.warning("Failed to decompress \(fileName)")
                        }
                    } else {
                        SDKLogger.archive.warning("Unsupported compression method \(compressionMethod) for \(fileName)")
                    }

                    fileCount += 1
                }

                offset = dataEnd
                progressHandler?(Double(offset) / Double(totalSize))
            } else if sig0 == 0x50 && sig1 == 0x4b && sig2 == 0x01 && sig3 == 0x02 {
                // Central directory - we're done with files
                break
            } else if sig0 == 0x50 && sig1 == 0x4b && sig2 == 0x05 && sig3 == 0x06 {
                // End of central directory
                break
            } else {
                offset += 1
            }
        }

        SDKLogger.archive.info("Extracted \(fileCount) files from zip")
        progressHandler?(1.0)
    }

    /// Decompress deflate data for ZIP files
    private static func decompressDeflateDataForZip(_ data: Data, uncompressedSize: Int) -> Data? {
        var destinationBufferSize = max(uncompressedSize, data.count * 4)
        var decompressedData = Data(count: destinationBufferSize)

        let decompressedSize = data.withUnsafeBytes { (srcPtr: UnsafeRawBufferPointer) -> Int in
            guard let sourceAddress = srcPtr.baseAddress else { return 0 }

            return decompressedData.withUnsafeMutableBytes { (destPtr: UnsafeMutableRawBufferPointer) -> Int in
                guard let destAddress = destPtr.baseAddress else { return 0 }

                // Use COMPRESSION_ZLIB for raw deflate
                return compression_decode_buffer(
                    destAddress.assumingMemoryBound(to: UInt8.self),
                    destinationBufferSize,
                    sourceAddress.assumingMemoryBound(to: UInt8.self),
                    data.count,
                    nil,
                    COMPRESSION_ZLIB
                )
            }
        }

        guard decompressedSize > 0 else { return nil }
        decompressedData.count = decompressedSize
        return decompressedData
    }

    // MARK: - TAR Extraction (Pure Swift)

    /// Extract tar data to destination directory
    private static func extractTarData(
        _ tarData: Data,
        to destinationURL: URL,
        progressHandler: ((Double) -> Void)?
    ) throws {
        // Create destination directory
        try FileManager.default.createDirectory(at: destinationURL, withIntermediateDirectories: true)

        var offset = 0
        let totalSize = tarData.count
        var fileCount = 0

        while offset + 512 <= tarData.count {
            // Read tar header (512 bytes)
            let headerData = tarData.subdata(in: offset..<(offset + 512))

            // Check for end of archive (two consecutive zero blocks)
            if headerData.allSatisfy({ $0 == 0 }) {
                break
            }

            // Parse header
            let nameData = headerData.subdata(in: 0..<100)
            let sizeData = headerData.subdata(in: 124..<136)
            let typeFlag = headerData[156]
            let prefixData = headerData.subdata(in: 345..<500)

            // Get file name
            let name = extractNullTerminatedString(from: nameData)
            let prefix = extractNullTerminatedString(from: prefixData)
            let fullName = prefix.isEmpty ? name : "\(prefix)/\(name)"

            // Skip if name is empty or is macOS resource fork
            guard !fullName.isEmpty, !fullName.hasPrefix("._") else {
                offset += 512
                continue
            }

            // Parse file size (octal)
            let sizeString = extractNullTerminatedString(from: sizeData).trimmingCharacters(in: .whitespaces)
            let fileSize = Int(sizeString, radix: 8) ?? 0

            offset += 512 // Move past header

            let filePath = destinationURL.appendingPathComponent(fullName)

            // Handle different entry types
            if typeFlag == 0x35 || (typeFlag == 0x30 && fullName.hasSuffix("/")) { // Directory
                try FileManager.default.createDirectory(at: filePath, withIntermediateDirectories: true)
            } else if typeFlag == 0x30 || typeFlag == 0 { // Regular file
                // Ensure parent directory exists
                try FileManager.default.createDirectory(at: filePath.deletingLastPathComponent(), withIntermediateDirectories: true)

                // Extract file data
                if fileSize > 0 && offset + fileSize <= tarData.count {
                    let fileData = tarData.subdata(in: offset..<(offset + fileSize))
                    try fileData.write(to: filePath)
                } else {
                    // Create empty file
                    FileManager.default.createFile(atPath: filePath.path, contents: nil)
                }
                fileCount += 1
            } else if typeFlag == 0x32 { // Symbolic link
                let linkName = extractNullTerminatedString(from: headerData.subdata(in: 157..<257))
                if !linkName.isEmpty {
                    try FileManager.default.createDirectory(at: filePath.deletingLastPathComponent(), withIntermediateDirectories: true)
                    try? FileManager.default.createSymbolicLink(atPath: filePath.path, withDestinationPath: linkName)
                }
            }

            // Move to next entry (file data + padding to 512-byte boundary)
            offset += fileSize
            let padding = (512 - (fileSize % 512)) % 512
            offset += padding

            // Report progress
            progressHandler?(Double(offset) / Double(totalSize))
        }

        SDKLogger.archive.info("Extracted \(fileCount) files")
    }

    // MARK: - Helpers

    private static func extractNullTerminatedString(from data: Data) -> String {
        if let nullIndex = data.firstIndex(of: 0) {
            return String(data: data.subdata(in: 0..<nullIndex), encoding: .utf8) ?? ""
        }
        return String(data: data, encoding: .utf8) ?? ""
    }

    private static func formatBytes(_ bytes: Int) -> String {
        if bytes < 1024 {
            return "\(bytes) B"
        } else if bytes < 1024 * 1024 {
            return String(format: "%.1f KB", Double(bytes) / 1024)
        } else if bytes < 1024 * 1024 * 1024 {
            return String(format: "%.1f MB", Double(bytes) / (1024 * 1024))
        } else {
            return String(format: "%.2f GB", Double(bytes) / (1024 * 1024 * 1024))
        }
    }
}

File: ios/runanywhere-sdks-latest/sdk/runanywhere-react-native/packages/core/android/src/main/java/com/margelo/nitro/runanywhere/ArchiveUtility.kt
/**
 * ArchiveUtility.kt
 *
 * Native archive extraction utility for Android.
 * Uses Apache Commons Compress for robust tar.gz extraction (streaming, memory-efficient).
 * Uses Java's native ZipInputStream for zip extraction.
 *
 * Mirrors the implementation from:
 * sdk/runanywhere-swift/Sources/RunAnywhere/Infrastructure/Download/Utilities/ArchiveUtility.swift
 *
 * Supports: tar.gz, zip
 * Note: All models should use tar.gz from RunanywhereAI/sherpa-onnx fork for best performance
 */

package com.margelo.nitro.runanywhere

import org.apache.commons.compress.archivers.tar.TarArchiveEntry
import org.apache.commons.compress.archivers.tar.TarArchiveInputStream
import org.apache.commons.compress.compressors.gzip.GzipCompressorInputStream
import java.io.BufferedInputStream
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.util.zip.ZipInputStream

/**
 * Utility for handling archive extraction on Android
 */
object ArchiveUtility {
    private val logger = SDKLogger.archive

    /**
     * Extract an archive to a destination directory
     * @param archivePath Path to the archive file
     * @param destinationPath Destination directory path
     * @return true if extraction succeeded
     */
    @JvmStatic
    fun extract(archivePath: String, destinationPath: String): Boolean {
        logger.info("extract() called: $archivePath -> $destinationPath")
        return try {
            extractArchive(archivePath, destinationPath)
            logger.info("extract() succeeded")
            true
        } catch (e: Exception) {
            logger.logError(e, "Extraction failed")
            false
        }
    }

    /**
     * Extract an archive to a destination directory (throwing version)
     */
    @Throws(Exception::class)
    fun extractArchive(
        archivePath: String,
        destinationPath: String,
        progressHandler: ((Double) -> Unit)? = null
    ) {
        val archiveFile = File(archivePath)
        val destinationDir = File(destinationPath)

        if (!archiveFile.exists()) {
            throw Exception("Archive not found: $archivePath")
        }

        // Detect archive type by magic bytes (more reliable than file extension)
        val archiveType = detectArchiveTypeByMagicBytes(archiveFile)
        logger.info("Detected archive type: $archiveType for: $archivePath")

        when (archiveType) {
            ArchiveType.GZIP -> {
                extractTarGz(archiveFile, destinationDir, progressHandler)
            }
            ArchiveType.ZIP -> {
                extractZip(archiveFile, destinationDir, progressHandler)
            }
            ArchiveType.BZIP2 -> {
                throw Exception("tar.bz2 not supported. Use tar.gz from RunanywhereAI/sherpa-onnx fork.")
            }
            ArchiveType.XZ -> {
                throw Exception("tar.xz not supported. Use tar.gz from RunanywhereAI/sherpa-onnx fork.")
            }
            ArchiveType.UNKNOWN -> {
                // Fallback to file extension check
                val lowercased = archivePath.lowercase()
                when {
                    lowercased.endsWith(".tar.gz") || lowercased.endsWith(".tgz") -> {
                        extractTarGz(archiveFile, destinationDir, progressHandler)
                    }
                    lowercased.endsWith(".zip") -> {
                        extractZip(archiveFile, destinationDir, progressHandler)
                    }
                    else -> {
                        throw Exception("Unknown archive format: $archivePath")
                    }
                }
            }
        }
    }

    /**
     * Archive type detected by magic bytes
     */
    private enum class ArchiveType {
        GZIP, ZIP, BZIP2, XZ, UNKNOWN
    }

    /**
     * Detect archive type by reading magic bytes from file header
     */
    private fun detectArchiveTypeByMagicBytes(file: File): ArchiveType {
        return try {
            FileInputStream(file).use { fis ->
                val header = ByteArray(6)
                val bytesRead = fis.read(header)
                if (bytesRead < 2) return ArchiveType.UNKNOWN

                // Check for gzip: 0x1f 0x8b
                if (header[0] == 0x1f.toByte() && header[1] == 0x8b.toByte()) {
                    return ArchiveType.GZIP
                }

                // Check for zip: 0x50 0x4b 0x03 0x04 ("PK\x03\x04")
                if (bytesRead >= 4 &&
                    header[0] == 0x50.toByte() && header[1] == 0x4b.toByte() &&
                    header[2] == 0x03.toByte() && header[3] == 0x04.toByte()) {
                    return ArchiveType.ZIP
                }

                // Check for bzip2: 0x42 0x5a ("BZ")
                if (header[0] == 0x42.toByte() && header[1] == 0x5a.toByte()) {
                    return ArchiveType.BZIP2
                }

                // Check for xz: 0xfd 0x37 0x7a 0x58 0x5a 0x00
                if (bytesRead >= 6 &&
                    header[0] == 0xfd.toByte() && header[1] == 0x37.toByte() &&
                    header[2] == 0x7a.toByte() && header[3] == 0x58.toByte() &&
                    header[4] == 0x5a.toByte() && header[5] == 0x00.toByte()) {
                    return ArchiveType.XZ
                }

                ArchiveType.UNKNOWN
            }
        } catch (e: Exception) {
            logger.error("Failed to detect archive type: ${e.message}")
            ArchiveType.UNKNOWN
        }
    }

    // MARK: - tar.gz Extraction

    /**
     * Extract a tar.gz archive using Apache Commons Compress (streaming, memory-efficient)
     * This approach doesn't load the entire file into memory.
     */
    private fun extractTarGz(
        sourceFile: File,
        destinationDir: File,
        progressHandler: ((Double) -> Unit)?
    ) {
        val startTime = System.currentTimeMillis()
        logger.info("Extracting tar.gz: ${sourceFile.name} (size: ${formatBytes(sourceFile.length())})")
        progressHandler?.invoke(0.0)

        destinationDir.mkdirs()
        var fileCount = 0
        val totalSize = sourceFile.length()
        var bytesRead = 0L

        try {
            // Use Apache Commons Compress for streaming tar.gz extraction
            FileInputStream(sourceFile).use { fis ->
                BufferedInputStream(fis).use { bis ->
                    GzipCompressorInputStream(bis).use { gzis ->
                        TarArchiveInputStream(gzis).use { tarIn ->
                            var entry: TarArchiveEntry? = tarIn.nextTarEntry
                            while (entry != null) {
                                val name = entry.name

                                // Skip macOS resource forks and empty names
                                if (name.isEmpty() || name.startsWith("._") || name.startsWith("./._")) {
                                    entry = tarIn.nextTarEntry
                                    continue
                                }

                                val outputFile = File(destinationDir, name)

                                // Security check - prevent zip slip attack
                                val destDirPath = destinationDir.canonicalPath
                                val outputFilePath = outputFile.canonicalPath
                                if (!outputFilePath.startsWith(destDirPath + File.separator) &&
                                    outputFilePath != destDirPath) {
                                    logger.warning("Skipping entry outside destination: $name")
                                    entry = tarIn.nextTarEntry
                                    continue
                                }

                                if (entry.isDirectory) {
                                    outputFile.mkdirs()
                                } else {
                                    // Create parent directories
                                    outputFile.parentFile?.mkdirs()

                                    // Extract file
                                    FileOutputStream(outputFile).use { fos ->
                                        val buffer = ByteArray(8192)
                                        var len: Int
                                        while (tarIn.read(buffer).also { len = it } != -1) {
                                            fos.write(buffer, 0, len)
                                            bytesRead += len
                                        }
                                    }
                                    fileCount++

                                    // Log progress for large files
                                    if (fileCount % 10 == 0) {
                                        logger.debug("Extracted $fileCount files...")
                                    }
                                }

                                // Report progress (estimate based on compressed bytes)
                                val progress = (bytesRead.toDouble() / (totalSize * 3)).coerceAtMost(0.95)
                                progressHandler?.invoke(progress)

                                entry = tarIn.nextTarEntry
                            }
                        }
                    }
                }
            }

            val totalTime = System.currentTimeMillis() - startTime
            logger.info("Extracted $fileCount files in ${totalTime}ms")
            progressHandler?.invoke(1.0)
        } catch (e: Exception) {
            logger.logError(e, "tar.gz extraction failed")
            throw e
        }
    }

    // MARK: - ZIP Extraction

    /**
     * Extract a zip archive using Java's native ZipInputStream
     */
    private fun extractZip(
        sourceFile: File,
        destinationDir: File,
        progressHandler: ((Double) -> Unit)?
    ) {
        logger.info("Extracting zip: ${sourceFile.name}")
        progressHandler?.invoke(0.0)

        destinationDir.mkdirs()

        var fileCount = 0
        ZipInputStream(BufferedInputStream(FileInputStream(sourceFile))).use { zis ->
            var entry = zis.nextEntry
            while (entry != null) {
                val fileName = entry.name
                val newFile = File(destinationDir, fileName)

                // Security check - prevent zip slip attack
                val destDirPath = destinationDir.canonicalPath
                val newFilePath = newFile.canonicalPath
                if (!newFilePath.startsWith(destDirPath + File.separator)) {
                    throw Exception("Entry is outside of the target dir: $fileName")
                }

                if (entry.isDirectory) {
                    newFile.mkdirs()
                } else {
                    // Create parent directories
                    newFile.parentFile?.mkdirs()

                    // Write file
                    FileOutputStream(newFile).use { fos ->
                        val buffer = ByteArray(8192)
                        var len: Int
                        while (zis.read(buffer).also { len = it } != -1) {
                            fos.write(buffer, 0, len)
                        }
                    }
                    fileCount++
                }

                zis.closeEntry()
                entry = zis.nextEntry
            }
        }

        logger.info("Extracted $fileCount files from zip")
        progressHandler?.invoke(1.0)
    }

    // MARK: - Helpers

    private fun formatBytes(bytes: Long): String {
        return when {
            bytes < 1024 -> "$bytes B"
            bytes < 1024 * 1024 -> String.format("%.1f KB", bytes / 1024.0)
            bytes < 1024 * 1024 * 1024 -> String.format("%.1f MB", bytes / (1024.0 * 1024))
            else -> String.format("%.2f GB", bytes / (1024.0 * 1024 * 1024))
        }
    }
}
